### PR TITLE
Implement a basic paging

### PR DESCRIPTION
**Pull request by Jasopaum**
## Fixes
Fixes https://github.com/arkhn/warehouse-api/issues/29

### CODE REVIEW

File: src/fhirpath/engine/es/__init__.py [hunk 1]
```diff
@@ -252,7 +252,9 @@ def process_raw_result(self, rawresult, selects, query_type):
             source_filters = self._get_source_filters(selects)
 
         result = EngineResult(
-            header=EngineResultHeader(total=total), body=EngineResultBody()
+            header=EngineResultHeader(total=total),
+            body=EngineResultBody(),
+            scroll_id=rawresult["_scroll_id"],
```
Comment (simonvadee) [inline src/fhirpath/engine/es/__init__.py]: .get()

File: src/fhirpath/engine/es/__init__.py [hunk 2]
```diff
@@ -261,24 +263,27 @@ def process_raw_result(self, rawresult, selects, query_type):
         if query_type != EngineQueryType.COUNT:
             self.extract_hits(source_filters, rawresult["hits"]["hits"], result.body)
 
-        if "_scroll_id" in rawresult and result.header.total > len(
-            rawresult["hits"]["hits"]
-        ):
-            # we need to fetch all!
-            consumed = len(rawresult["hits"]["hits"])
+        # TODO add link to next page in Result bundle
```
Comment (simonvadee) [inline src/fhirpath/engine/es/__init__.py]: rm

File: src/fhirpath/engine/es/__init__.py [hunk 3]
```diff
@@ -178,7 +178,9 @@ def _execute(self, query, unrestricted, query_type):
             calculate_field_index_name=self.calculate_field_index_name,
             get_mapping=self.get_mapping,
         )
-        if query_type == EngineQueryType.DML:
+        if query._scroll_id:
```
Comment (simonvadee) [inline src/fhirpath/engine/es/__init__.py]: should we use the enum EngineQueryType instead ? Even if it's not exactly the same as EngineQueryType.COUNT I think a query of type SCROLL will be handled in a specific way (just like _count) so it would make sense imo

File: src/fhirpath/engine/es/__init__.py (hunk 1 above)
Comment (simonvadee) [reply src/fhirpath/engine/es/__init__.py]: do we always have a _scroll_id in the raw result ?

File: src/fhirpath/engine/es/__init__.py [hunk 4]
```diff
@@ -261,36 +265,17 @@ def process_raw_result(self, rawresult, selects, query_type):
...
-                if len(raw_res["hits"]["hits"]) == 0:
-                    break
-
-                self.extract_hits(source_filters, raw_res["hits"]["hits"], result.body)
-
-                consumed += len(raw_res["hits"]["hits"])
-
-                if result.header.total <= consumed:
-                    break
-
```
Comment (simonvadee) [inline src/fhirpath/engine/es/__init__.py:340]: 🙌

File: src/fhirpath/search.py [hunk 5]
```diff
@@ -340,6 +341,21 @@ def __init__(self, context: SearchContext, query_string=None, params=None):
         if additional_resource_types:
             self.context.augment_with_types(additional_resource_types)
 
+        self.query_string = self.build_query_string_from_params(
+            query_string, all_params
+        )
```
Comment (simonvadee) [inline src/fhirpath/search.py]: only use all_params, which is always a filled MultiDict, and do it before prepare_params

File: src/fhirpath/search.py [hunk 6]
```diff
@@ -340,6 +341,21 @@ def __init__(self, context: SearchContext, query_string=None, params=None):
...
+
+    def build_query_string_from_params(self, query_string, params):
+        if query_string:
+            if re.search(url_type, query_string):
+                # Remove _type if it exists
+                query_string = re.sub(url_type, r"\1\2", query_string)
+            return f"{query_string}&_type={','.join(self.context.resource_types)}"
+        else:
```
Comment (simonvadee) [inline src/fhirpath/search.py]: remove all this

File: src/fhirpath/search.py [hunk 7]
```diff
@@ -1644,19 +1665,23 @@ def attach_limit_terms(self, builder):
                 offset = (current_page - 1) * self.result_params["_count"]
         return builder.limit(self.result_params["_count"], offset)
 
+    def attach_scroll_id(self, builder):
+        """ """
+        return builder.set_scroll_id(self.result_params.get("_scroll_id"))
```
Comment (simonvadee) [inline src/fhirpath/search.py]: you should add a check like:
```
        if "_scroll_id" not in self.result_params:
            return builder
```

File: src/fhirpath/search.py [hunk 8]
```diff
@@ -1644,19 +1665,23 @@ def attach_limit_terms(self, builder):
...
+
     def response(self, result, includes, as_json):
         """ """
         return self.context.engine.wrapped_with_bundle(
-            result, includes=includes, as_json=as_json
+            result, query_params=self.query_string, includes=includes, as_json=as_json
```
Comment (simonvadee) [inline src/fhirpath/search.py]: naming should be consistent: chose between query_string and query_params

File: src/fhirpath/utils.py [hunk 9]
```diff
@@ -584,48 +584,15 @@ def attach_entry(self, result, mode="match"):
...
-
-            if _current_offset < last_offset and last_offset > 0:
-
-                url_params["search-offset"] = last_offset
-                container.append(self.make_link("last", url, url_params))
-
-        self.data["link"] = container
+        links = [self.make_link("self", url)]
+
+        # let's pagitionation here
```
Comment (simonvadee) [inline src/fhirpath/utils.py]: have you kept this comment because it made you laugh ?

File: tests/_utils.py [hunk 10]
```diff
@@ -65,7 +65,7 @@ def get_mapping(self, resource_type):
         mapping = fhir_resource_mapping(resource_type)
         return mapping
 
-    def current_url(self):
+    def current_url(self, query_params):
```
Comment (simonvadee) [inline tests/_utils.py:69]: consistent naming

File: src/fhirpath/search.py [hunk 11]
```diff
@@ -1644,19 +1665,23 @@ def attach_limit_terms(self, builder):
...
 
     def __call__(self, as_json=False):
         """ """
-
         # TODO: chaining
 
         # reverse chaining (_has)
-        if self.result_params.get("_has"):
+        # TODO reverse chaining won't work together with paging
+        if "_has" in self.result_params and "_scroll_id" not in self.result_params:
```
Comment (simonvadee) [inline src/fhirpath/search.py:1677]: maybe we can use query_type.scroll ? maybe we should build the query before this if

**Review by simonvadee — CHANGES_REQUESTED** [review]

File: src/fhirpath/engine/es/__init__.py (hunk 1 above)
Comment (Jasopaum) [reply src/fhirpath/engine/es/__init__.py]: Now that we tell ES that we may want to scroll, it should always be there. But I think it's still better to use get

File: src/fhirpath/engine/es/__init__.py (hunk 3 above)
Comment (Jasopaum) [reply src/fhirpath/engine/es/__init__.py]: To be able to switch on a query type, I think I'd need to use a new method instead of `fetchall` in which I'd pass a new type (this is how `count` works).
What do you think?

Comment (simonvadee) [reply src/fhirpath/engine/es/__init__.py]: let's add a query.scroll method and introduce a new query_type

File: src/fhirpath/engine/base.py [hunk 12]
```diff
@@ -101,7 +101,7 @@ class EngineResult(object):
     body: EngineResultBody
 
     def __init__(
-        self, header: EngineResultHeader, body: EngineResultBody, scroll_id: str
+        self, header: EngineResultHeader, body: EngineResultBody, scroll_id: str = ""
```
Comment (simonvadee) [inline src/fhirpath/engine/base.py]: shouldn't the default scroll_id be None ? I am not a huge fan of empty strings for default "nullable" values

Comment (simonvadee) [reply src/fhirpath/engine/base.py]: oh, is it because URL.build does not accept None values ?

File: src/fhirpath/search.py [hunk 13]
```diff
@@ -351,6 +352,8 @@ def __init__(self, context: SearchContext, query_string=None, params=None):
             self.context.augment_with_types(additional_resource_types)
 
     def build_query_params_string(self, params):
+        # Otherwise, an expection is raised for None values
+        params = {k: v or "" for k, v in params.items()}
```
Comment (simonvadee) [inline src/fhirpath/search.py]: Do we want to keep params that don't have a value when building the url ?

File: src/fhirpath/search.py [hunk 14]
```diff
@@ -1700,8 +1703,10 @@ def __call__(self, as_json=False):
         self.main_query = self.build()
 
         # TODO handle count with _includes
-        if self.result_params.get("_summary") == "count":
+        if self.result_params.get("_summary") == "count" or self.result_params.get("_count") == 0:
```
Comment (simonvadee) [inline src/fhirpath/search.py]: why did you add this condition here ?

Comment (Jasopaum) [reply src/fhirpath/search.py]: I didn't get why I needed to add is neither... I'm not sure why but the test related to `_count=0` was failing because of my changes.
How was this done before? Where did we tell that `_count=0` is equivalent to `_summary=count`?

File: src/fhirpath/search.py (hunk 13 above)
Comment (Jasopaum) [reply src/fhirpath/search.py]: Hum, maybe not, indeed!

File: src/fhirpath/engine/base.py (hunk 12 above)
Comment (Jasopaum) [reply src/fhirpath/engine/base.py]: Ironic for a gopher, right? 😉 
I'll put None as default, I don't think it'll break anything

Comment (simonvadee) [reply src/fhirpath/engine/base.py]: *str ftw

File: src/fhirpath/search.py (hunk 14 above)
Comment (simonvadee) [reply src/fhirpath/search.py]: I don't have a fckin clue...

**Review by simonvadee — APPROVED** [review]
